fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AdjustPreAggStatus.java

```diff
@@ -0,0 +1,704 @@
+                // only scan.
+                logicalOlapScan().when(LogicalOlapScan::isPreAggStatusUnSet)
+                        .thenApplyNoThrow(ctx -> {
+                            LogicalOlapScan scan = ctx.root;
+                            List<AggregateFunction> aggregateFunctions = ImmutableList.of();
+                            List<Expression> groupByExpressions = ImmutableList.of();
+                            Set<Expression> predicates = ImmutableSet.of();
+                            PreAggStatus preAggStatus = checkPreAggStatus(scan, predicates,
+                                    aggregateFunctions, groupByExpressions);
+                            return scan.withPreAggStatus(preAggStatus);
+                        }).toRule(RuleType.PREAGG_STATUS_SCAN));
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Set pre-aggregation status.
+    ///////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Do aggregate function extraction and replace aggregate function's input slots by underlying project.
+     * <p>
+     * 1. extract aggregate functions in aggregate plan.
+     * <p>
+     * 2. replace aggregate function's input slot by underlying project expression if project is present.
+     * <p>
+     * For example:
+     * <pre>
+     * input arguments:
+     * agg: Aggregate(sum(v) as sum_value)
+     * underlying project: Project(a + b as v)
+     *
+     * output:
+     * sum(a + b)
+     * </pre>
+     */
+    private List<AggregateFunction> extractAggFunctionAndReplaceSlot(LogicalAggregate<?> agg,
+            Optional<LogicalProject<?>> project) {
+        Optional<Map<Slot, Expression>> slotToProducerOpt =
+                project.map(Project::getAliasToProducer);
+        return agg.getOutputExpressions().stream()
+                // extract aggregate functions.
+                .flatMap(e -> e.<Set<AggregateFunction>>collect(AggregateFunction.class::isInstance)
+                        .stream())
+                // replace aggregate function's input slot by its producing expression.
+                .map(expr -> slotToProducerOpt
+                        .map(slotToExpressions -> (AggregateFunction) ExpressionUtils.replace(expr,
+                                slotToExpressions))
+                        .orElse(expr))
+                .collect(Collectors.toList());
+    }
+
+    private PreAggStatus checkPreAggStatus(LogicalOlapScan olapScan, Set<Expression> predicates,
+            List<AggregateFunction> aggregateFuncs, List<Expression> groupingExprs) {
+        long selectIndexId = olapScan.getSelectedIndexId();
+        MaterializedIndexMeta meta = olapScan.getTable().getIndexMetaByIndexId(selectIndexId);
+        if (meta.getKeysType() == KeysType.DUP_KEYS || (meta.getKeysType() == KeysType.UNIQUE_KEYS
+                && olapScan.getTable().getEnableUniqueKeyMergeOnWrite())) {
+            return PreAggStatus.on();
+        }
```

Comment:
let it be a short path method before we do extract and normalize agg function? for most case, this will work well